selfservice/strategy/oidc/strategy.go

```diff
@@ -457,6 +457,16 @@ func (s *Strategy) handleError(w http.ResponseWriter, r *http.Request, f flow.Fl
 		// Reset all nodes to not confuse users.
 		// This is kinda hacky and will probably need to be updated at some point.
 
+		if errors.Is(err, registration.ErrDuplicateCredentials) {
+			rf.UI.Messages.Add(text.NewErrorValidationDuplicateCredentialsOnOIDCLink())
+			lf, err := s.registrationToLogin(w, r, rf, provider)
+			if err != nil {
+				return err
+			}
+			x.AcceptToRedirectOrJSON(w, r, s.d.Writer(), lf, lf.AppendTo(s.d.Config().SelfServiceFlowLoginUI(r.Context())).String())
+			return nil
```

Comment:
The `x.AcceptToRedirectJSON` doesn't return anything. 

It redirects or writes json back to the browser.

the return nil just prevents the rest of the function from executing. In this case we don't want to propagate the error further since we want to return a login flow with the error message attached to the login flow.

Reply:
We are always calling this function as part of the return part of functions, which return `someValue, error`. For example here https://github.com/ory/kratos/pull/3151/files/9b1b977a8bbc43e86687bb631fc46ac4f3d08241.

In that example, if you now return `nil` from `handleError`, the returned values of `processRegistration` will be `nil, nil`. And that is an issue, because the caller of `processRegistration` ([here](https://github.com/ory/kratos/blob/9c3bfe3d1b3fe69923589e96ba55d919efcc3b3c/selfservice/strategy/oidc/strategy_login.go#L122) for example) will check whether `err != nil`, which is `false` (because you returned `nil` here) and then assume that the `*login.Flow` is a valid object and `processRegistration` completed successfully. It will then go on and reference the `nil` pointer (`*login.Flow`), which will panic.

I would really like to see a test that executes this code path. Did you already write one?

Reply:
I have a test for this though.

Reply:
Please see https://github.com/ory/kratos/pull/3151/files#diff-03d9986942bad676c4567865d479f219a54bb633d7ccc9eed5820703e3d296a3

and an e2e test https://github.com/ory/kratos/pull/3151/files#diff-2695d9d779ed70b5b6a2d9cc5111dcac495b41bfc8353c0e0a50154ff787f7a6

Remember this PR changes the behaviour of the registration flow with `oidc` when the account already exists.
Instead of just giving back an error code and message, it responds with a new Login flow and the error message inside the `login` flow.

Reply:
I think you should return `ErrAbortFlow` here. This prevents further calls in case something changes at some point.